File: 백준/UnionFind/1976.js
const fs = require("fs");
const inputs = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const n = Number(inputs[0]);
const graph = inputs
  .slice(2, -1)
  .map((e) => e.split(" ").map((e) => Number(e)));
const plan = inputs
  .slice(-1)[0]
  .split(" ")
  .map((e) => Number(e));

const parent = Array.from({ length: n + 1 }, (_, i) => i);

const getParent = (a, parent) => {
  if (a === parent[a]) {
    return a;
  }

  return (parent[a] = getParent(parent[a], parent));
};

const union = (a, b, parent) => {
  let x = getParent(a, parent);
  let y = getParent(b, parent);

  if (x < y) {
    parent[y] = x;
  } else {
    parent[x] = y;
  }
};

graph.forEach((row, i) => {
  row.forEach((col, j) => {
    if (col) {
      union(i + 1, j + 1, parent);
    }
  });
});

for (let i = 1; i <= n; i++) {
  getParent(i, parent);
}

let num = new Set(plan.map((p) => parent[p])).size;

console.log(num === 1 ? "YES" : "NO");

File: 백준/UnionFind/1976.py
import sys
n = int(sys.stdin.readline().strip())
m = int(sys.stdin.readline().strip())
graph = [list(map(int, sys.stdin.readline().strip().split())) for _ in range(n)]
plan = list(map(lambda x: int(x) - 1, sys.stdin.readline().strip().split()))

parent = [i for i in range(n)]

def get_parent(i, parent):
    if i != parent[i]:
        parent[i] = get_parent(parent[i], parent)
        return parent[i]
    
    return i

def union(a, b, parent):
    parent_a = get_parent(a, parent)
    parent_b = get_parent(b, parent)

    if parent_a < parent_b:
        parent[parent_b] = parent_a
    else:
        parent[parent_a] = parent_b

for i in range(n):
    for j in range(n):
        if i <= j and graph[i][j] == 1:
            union(i, j, parent)

for i in range(n):
    get_parent(i, parent)        

prev = plan[0]
for i in range(1, m):
    curr = plan[i]

    if parent[prev] != parent[curr]:
        print("NO")
        break
    else:
        prev = curr
else:
    print("YES")

    